File: backend/app/ewity_client.py
"""Ewity API client with caching"""
import httpx
import json
from datetime import datetime
from typing import Optional, List, Dict, Any
from sqlalchemy.orm import Session
from .config import get_settings
from .cache import cache

settings = get_settings()


class EwityClient:
    """Client for Ewity POS API"""

    def __init__(self):
        self.base_url = settings.ewity_api_base_url
        self.headers = {"Authorization": f"Bearer {settings.ewity_api_token}"}

    async def _get(self, endpoint: str, params: Optional[Dict] = None) -> Dict[str, Any]:
        """Make GET request to Ewity API"""
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{self.base_url}{endpoint}",
                headers=self.headers,
                params=params,
                timeout=10.0
            )
            response.raise_for_status()
            return response.json()

    async def get_customer(self, customer_id: int) -> Optional[Dict[str, Any]]:
        """Get customer by ID with caching"""
        cache_key = f"customer:{customer_id}"

        # Check cache first
        cached = cache.get(cache_key)
        if cached:
            return cached

        # Fetch all customers and find the one we want
        # Note: Ewity doesn't have a single customer endpoint, so we need to search
        try:
            data = await self._get("/customers", params={"pageSize": 100})
            customers = data.get("data", [])

            # Search through customers
            for customer in customers:
                if customer.get("id") == customer_id:
                    cache.set(cache_key, customer)
                    return customer

            # If not found in first page, we might need pagination
            # For now, return None
            return None
        except Exception as e:
            print(f"Error fetching customer {customer_id}: {e}")
            return None

    async def search_customers(self, query: str, page: int = 1, db: Optional[Session] = None) -> Dict[str, Any]:
        """Search customers by name or phone from local database"""
        from .models import Customer

        try:
            if not db:
                print("Warning: No database session provided for search")
                return {"data": [], "pagination": {}}

            # Search in local database
            query_lower = query.lower()

            # Use SQLAlchemy to search
            results = db.query(Customer).filter(
                (Customer.name.ilike(f"%{query}%")) |
                (Customer.mobile.like(f"%{query}%"))
            ).all()

            # Convert to dict format
            filtered = []
            for customer in results:
                customer_dict = {
                    "id": customer.id,
                    "name": customer.name,
                    "mobile": customer.mobile,
                    "email": customer.email,
                    "address": customer.address,
                    "creditLimit": customer.credit_limit,
                    "totalSpent": customer.total_spent,
                    "outstandingBalance": customer.outstanding_balance
                }
                filtered.append(customer_dict)

            # Paginate the filtered results
            page_size = 20
            start_idx = (page - 1) * page_size
            end_idx = start_idx + page_size
            paginated_filtered = filtered[start_idx:end_idx]

            result = {
                "data": paginated_filtered,
                "pagination": {
                    "total": len(filtered),
                    "page": page,
                    "pageSize": page_size,
                    "totalPages": (len(filtered) + page_size - 1) // page_size
                }
            }

            return result
        except Exception as e:
            print(f"Error searching customers: {e}")
            return {"data": [], "pagination": {}}

    async def get_all_customers(self, page: int = 1) -> Dict[str, Any]:
        """Get all customers with caching"""
        cache_key = f"customers:page:{page}"

        cached = cache.get(cache_key)
        if cached:
            return cached

        try:
            data = await self._get("/customers", params={"page": page, "pageSize": 20})
            cache.set(cache_key, data)
            return data
        except Exception as e:
            print(f"Error fetching customers: {e}")
            return {"data": [], "pagination": {}}

    async def sync_all_customers_to_db(self, db: Session) -> Dict[str, Any]:
        """Fetch all customers from Ewity and sync to local database"""
        from .models import Customer

        try:
            print("🔄 Syncing customers from Ewity API...")

            synced_count = 0
            updated_count = 0
            page = 1
            total_pages = 1

            # Fetch all pages
            while page <= total_pages:
                print(f"  Fetching page {page}...")

                # Fetch customers page by page (API always returns 20 per page regardless of pageSize)
                data = await self._get("/customers", params={"page": page})
                customers = data.get("data", [])
                pagination = data.get("pagination", {})

                # Update total pages from first response
                if page == 1:
                    # API uses 'lastPage' not 'totalPages'
                    total_pages = pagination.get("lastPage", 1)
                    total_customers = pagination.get("total", 0)
                    print(f"  Found {total_pages} pages ({total_customers} total customers)")

                # Process customers from this page
                for customer_data in customers:
                    customer_id = customer_data.get("id")
                    if not customer_id:
                        continue

                    # Check if customer exists
                    existing = db.query(Customer).filter(Customer.id == customer_id).first()

                    customer_obj = existing or Customer(id=customer_id)
                    customer_obj.name = customer_data.get("name")
                    customer_obj.mobile = customer_data.get("mobile")
                    customer_obj.email = customer_data.get("email")
                    customer_obj.address = customer_data.get("address")
                    customer_obj.credit_limit = customer_data.get("creditLimit")
                    customer_obj.total_spent = customer_data.get("totalSpent")
                    customer_obj.outstanding_balance = customer_data.get("outstandingBalance")
                    customer_obj.data = json.dumps(customer_data)
                    customer_obj.synced_at = datetime.utcnow()

                    if existing:
                        updated_count += 1
                    else:
                        db.add(customer_obj)
                        synced_count += 1

                # Commit after each page to avoid memory issues
                db.commit()
                print(f"  ✓ Processed page {page}/{total_pages}")

                page += 1

            total = synced_count + updated_count
            print(f"✓ Synced {total} customers ({synced_count} new, {updated_count} updated)")

            return {
                "success": True,
                "total": total,
                "new": synced_count,
                "updated": updated_count
            }

        except Exception as e:
            print(f"❌ Error syncing customers: {e}")
            db.rollback()
            return {
                "success": False,
                "error": str(e)
            }


# Global client instance
ewity_client = EwityClient()

File: backend/backend-nodejs/src/lib/ewity-client.ts
import { config } from '../config.js';
import { db } from './database.js';
import { cache } from './cache.js';
import type { EwityCustomer, EwityCustomersResponse, Customer } from '../types.js';

export class EwityClient {
  private baseUrl: string;
  private headers: Record<string, string>;

  constructor() {
    this.baseUrl = config.ewityApiBaseUrl;
    this.headers = {
      'Authorization': `Bearer ${config.ewityApiToken}`,
    };
  }

  private async get(endpoint: string, params?: Record<string, any>): Promise<any> {
    const url = new URL(`${this.baseUrl}${endpoint}`);
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        url.searchParams.append(key, String(value));
      });
    }

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers: this.headers,
    });

    if (!response.ok) {
      throw new Error(`Ewity API error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  }

  async getCustomer(customerId: string): Promise<EwityCustomer | null> {
    try {
      // Check local database first
      const customer = await db.getCustomerById(customerId);

      if (customer) {
        // Return customer data from local database
        return {
          id: customer.id,
          name: customer.name,
          mobile: customer.mobile,
          email: customer.email || undefined,
          address: customer.address || undefined,
          credit_limit: customer.credit_limit,
          creditLimit: customer.credit_limit,
          total_spent: customer.total_spent,
          totalSpent: customer.total_spent,
          total_outstanding: customer.outstanding_balance,
          outstandingBalance: customer.outstanding_balance,
        };
      }

      // If not in local database, fetch from API
      console.log(`Customer ${customerId} not in local DB, fetching from API...`);
      const data = await this.get('/customers', { page: 1 });
      const customers = data?.data || [];

      for (const customerData of customers) {
        if (customerData.id === customerId) {
          // Store in database for future use
          await db.upsertCustomer({
            id: customerId,
            name: customerData.name,
            mobile: customerData.mobile,
            email: customerData.email,
            address: customerData.address,
            credit_limit: customerData.credit_limit,
            total_spent: customerData.total_spent,
            outstanding_balance: customerData.total_outstanding,
            data: JSON.stringify(customerData),
          });
          return customerData;
        }
      }

      return null;
    } catch (error) {
      console.error(`Error fetching customer ${customerId}:`, error);
      return null;
    }
  }

  async searchCustomers(query: string, page: number = 1): Promise<EwityCustomersResponse> {
    try {
      // Search in local database
      const customers = await db.searchCustomers(query, 50);

      // Convert to API format
      const data = customers.map(customer => ({
        id: customer.id,
        name: customer.name,
        mobile: customer.mobile,
        email: customer.email || undefined,
        address: customer.address || undefined,
        creditLimit: customer.credit_limit,
        totalSpent: customer.total_spent,
        outstandingBalance: customer.outstanding_balance,
      }));

      // Simple pagination
      const pageSize = 20;
      const startIdx = (page - 1) * pageSize;
      const endIdx = startIdx + pageSize;
      const paginatedData = data.slice(startIdx, endIdx);

      return {
        customers: paginatedData,
        page,
        total_pages: Math.ceil(data.length / pageSize),
        total_count: data.length,
      };
    } catch (error) {
      console.error('Error searching customers:', error);
      return {
        customers: [],
        page,
        total_pages: 0,
        total_count: 0,
      };
    }
  }

  async getAllCustomers(page: number = 1): Promise<any> {
    const cacheKey = `customers:page:${page}`;

    // Check cache first
    const cached = await cache.get(cacheKey);
    if (cached) {
      return cached;
    }

    try {
      const data = await this.get('/customers', { page, pageSize: 20 });
      await cache.set(cacheKey, data, 300); // Cache for 5 minutes
      return data;
    } catch (error) {
      console.error('Error fetching customers:', error);
      return { data: [], pagination: {} };
    }
  }

  async syncAllCustomersToDb(): Promise<{ success: boolean; total?: number; new?: number; updated?: number; error?: string }> {
    try {
      console.log('🔄 Syncing customers from Ewity API...');

      let syncedCount = 0;
      let updatedCount = 0;
      let page = 1;
      let totalPages = 1;

      // Fetch all pages
      while (page <= totalPages) {
        console.log(`  Fetching page ${page}...`);

        const data = await this.get('/customers', { page });
        const customers = data?.data || [];
        const pagination = data?.pagination || {};

        // Update total pages from first response
        if (page === 1) {
          totalPages = pagination.lastPage || 1;
          const totalCustomers = pagination.total || 0;
          console.log(`  Found ${totalPages} pages (${totalCustomers} total customers)`);
        }

        // Process customers from this page
        for (const customerData of customers) {
          const customerId = customerData.id;
          if (!customerId) {
            continue;
          }

          // Check if customer exists
          const existing = await db.getCustomerById(customerId);

          await db.upsertCustomer({
            id: customerId,
            name: customerData.name,
            mobile: customerData.mobile,
            email: customerData.email,
            address: customerData.address,
            credit_limit: customerData.credit_limit,
            total_spent: customerData.total_spent,
            outstanding_balance: customerData.total_outstanding,
            data: JSON.stringify(customerData),
          });

          if (existing) {
            updatedCount++;
          } else {
            syncedCount++;
          }
        }

        console.log(`  ✓ Processed page ${page}/${totalPages}`);
        page++;
      }

      const total = syncedCount + updatedCount;
      console.log(`✓ Synced ${total} customers (${syncedCount} new, ${updatedCount} updated)`);

      return {
        success: true,
        total,
        new: syncedCount,
        updated: updatedCount,
      };
    } catch (error) {
      console.error('❌ Error syncing customers:', error);
      return {
        success: false,
        error: String(error),
      };
    }
  }
}

export const ewityClient = new EwityClient();
